File: October_2025/potd_04_10_2025.py
class Solution:
    def addOperators(self, s: str, target: int):
        res = []

        def dfs(idx, path, eval_val, prev):
            if idx == len(s):
                if eval_val == target:
                    res.append(path)
                return

            for i in range(idx, len(s)):
                if i != idx and s[idx] == '0':
                    break
                cur_str = s[idx:i+1]
                cur = int(cur_str)

                if idx == 0:
                    dfs(i+1, cur_str, cur, cur)
                else:
                    dfs(i+1, path + "+" + cur_str, eval_val + cur, cur)
                    dfs(i+1, path + "-" + cur_str, eval_val - cur, -cur)
                    dfs(i+1, path + "*" + cur_str, eval_val - prev + prev * cur, prev * cur)

        dfs(0, "", 0, 0)
        res.sort()
        return res


# Example
sol = Solution()
print(sol.addOperators("125", 7))  # Output: ['1*2+5', '12-5']

File: October_2025/potd_04_10_2025.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    vector<string> res;

    void dfs(string &s, long long target, int idx, string path, long long eval, long long prev)
    {
        if (idx == s.size())
        {
            if (eval == target)
                res.push_back(path);
            return;
        }

        for (int i = idx; i < s.size(); i++)
        {
            if (i != idx && s[idx] == '0')
                break; // skip leading zero numbers
            string curStr = s.substr(idx, i - idx + 1);
            long long cur = stoll(curStr);

            if (idx == 0)
            {
                dfs(s, target, i + 1, curStr, cur, cur);
            }
            else
            {
                dfs(s, target, i + 1, path + "+" + curStr, eval + cur, cur);
                dfs(s, target, i + 1, path + "-" + curStr, eval - cur, -cur);
                dfs(s, target, i + 1, path + "*" + curStr, eval - prev + prev * cur, prev * cur);
            }
        }
    }

    vector<string> addOperators(string s, int target)
    {
        res.clear();
        dfs(s, target, 0, "", 0, 0);
        sort(res.begin(), res.end());
        return res;
    }
};

int main()
{
    Solution sol;
    string s = "125";
    int target = 7;
    vector<string> ans = sol.addOperators(s, target);
    for (auto &expr : ans)
        cout << expr << " ";
}